hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java

```diff
@@ -317,33 +318,40 @@ protected HoodieWriteMetadata<O> compact(HoodieTable<?, I, ?, T> table, String c
     }
     compactionTimer = metrics.getCompactionCtx();
     HoodieWriteMetadata<T> writeMetadata = table.compact(context, compactionInstantTime);
-    HoodieWriteMetadata<O> compactionWriteMetadata = convertToOutputMetadata(writeMetadata);
+    HoodieWriteMetadata<T> processedWriteMetadata = processWriteMetadata(table, writeMetadata, compactionInstantTime);
+    HoodieWriteMetadata<O> compactionWriteMetadata = convertToOutputMetadata(processedWriteMetadata);
     if (shouldComplete) {
       commitCompaction(compactionInstantTime, compactionWriteMetadata, Option.of(table));
     }
     return compactionWriteMetadata;
   }
 
+  protected HoodieWriteMetadata<T> processWriteMetadata(HoodieTable table, HoodieWriteMetadata<T> writeMetadata, String instantTime) {
+    return writeMetadata;
+  }
+
   public void commitCompaction(String compactionInstantTime, HoodieWriteMetadata<O> compactionWriteMetadata, Option<HoodieTable> tableOpt) {
     // dereferencing the write dag for compaction for the first time.
-    List<HoodieWriteStat> writeStats = triggerWritesAndFetchWriteStats(compactionWriteMetadata);
+    Pair<List<HoodieWriteStat>, List<HoodieWriteStat>> dataTableAndMetadataTableHoodieWriteStats = triggerWritesAndFetchWriteStats(compactionWriteMetadata);
```

Comment:
I think it would be better for other devs to have a POJO here so it is clear which half of the pair is for the data table vs metadata table. If you disagree, can we at least introduce some local variables in the methods using the result that have data/metadata in the name so it is more clear what is passed to the other methods.

Reply:
sure. Adding TableWriteStatsHolder